NodeSelector, HardNodeAffinity and SoftNodeAffinity implementation

kevin-wangzefeng opened the pull request:
ref #18261 

This PR contains the implementation of:
1. NodeSelector, HardNodeAffinity and SoftNodeAffinity API types and validation
2. api helper to covert NodeSelector to labels.selector
3. Scheduling predicates update, supporting HardNodeAffinity
4. New scheduling priority algorithm take SoftNodeAffinity into account

kevin-wangzefeng commented:
@davidopp PTAL

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1323,6 +1324,74 @@ func ValidatePodSpec(spec *api.PodSpec, fldPath *field.Path) field.ErrorList {
+	return allErrs
+}
+
+// ValidatePodAffinity test that the specified Affinity field in PocSpec has valid data
```
s/PocSpec/PodSpec/

davidopp commented on pkg/labels/selector.go:
```diff
@@ -125,6 +129,15 @@ func NewRequirement(key string, op Operator, vals sets.String) (*Requirement, er
 		}
+	case GreaterThanOperator, LessThanOperator:
+		if len(vals) != 1 {
+			return nil, fmt.Errorf("for 'Gt', 'Lt' operators, only one single value is required")
```
s/only/exactly/
s/single//

davidopp commented on pkg/labels/selector.go:
```diff
@@ -162,6 +175,22 @@ func (r *Requirement) Matches(ls Labels) bool {
+
+		// there is only one strValue in r.strValues here, checked in NewRequirement()
+		var rValue float64
+		for strValue := range r.strValues {
```
it would still probably be a good idea to log an error if len(r.strValues) != 1

davidopp commented on pkg/labels/selector.go:
```diff
@@ -162,6 +175,22 @@ func (r *Requirement) Matches(ls Labels) bool {
+		var rValue float64
+		for strValue := range r.strValues {
+			// value is already checked in NewRequirement, not check again here.
+			rValue, _ = strconv.ParseFloat(strValue, 64)
```
check for error instead of discarding it

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -347,12 +347,50 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
 	return selector.PodSelectorMatches
 }
 
+// New versions of the scheduler will apply their scheduling predicates
```
This comment won't really make sense once this has been deployed for a while, and the full comment is in the design doc, so I would just replace it with "The pod can only schedule onto nodes that satisfy both hardNodeAffinity and nodeSelector."

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -347,12 +347,50 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+		selector := labels.SelectorFromSet(pod.Spec.NodeSelector)
+		nodeSelectorMatches = selector.Matches(labels.Set(node.Labels))
+	}
+
```
You could do
```
if (!nodeSelectorMatches)
     return false
```
here since you know the function can never return true in that case.

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -347,12 +347,50 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+	if pod.Spec.Affinity != nil && pod.Spec.Affinity.HardNodeAffinity != nil {
+		if len(pod.Spec.Affinity.HardNodeAffinity.NodeSelectorTerms) == 0 {
+			glog.V(10).Infof("No element in HardNodeAffinity.NodeSelectorTerms of Pod %+v, will match no labels of Node %+v.", podName(pod), node)
+			nodeAffinityMatches = labels.Nothing().Matches(labels.Set(node.Labels))
```
You might want to add a comment that this is guaranteed to be false because labels.Nothing().Matches() always returns false: https://github.com/kubernetes/kubernetes/blob/master/pkg/labels/selector.go#L51

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -347,12 +347,50 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+		}
+	}
+
+	return nodeSelectorMatches && nodeAffinityMatches
```
If you make the change I mentioned earlier, this could just become
```
return nodeAffinityMatches
```

davidopp commented:
@bgrant0607 should `Affinity` and the related types (like `NodeSelector`) go in v1 or in extensions/v1beta1? If we do the latter, then the v1 `PodSpec` would have a field that is a pointer to a type that lives in v1beta1. The situation we have here is a little different from when we added `LabelSelector`, since in that case all the types that used `LabelSelector` we put in v1beta1. In previous discussions we decided we didn't need to create a v1beta1.PodSpec, and would just AND the old nodeSelector map with the new Affinity stuff, so `PodSpec` will be in v1.

davidopp commented:
Please add some test cases to `test/e2e/scheduler_predicates.go` to test this feature.

kevin-wangzefeng commented:
Yes, definitely.

timothysc commented on api/swagger-spec/v1.json:
```diff
@@ -15011,6 +15015,95 @@
      }
     }
    },
+   "v1.Affinity": {
```
are you sure you want to go straight to v1 on 1st cut vs through extensions?  Changing will be difficult then...

kevin-wangzefeng replied:
Thanks for reminding, we do need more discussion about this.

I just copy @davidopp's prior [comment](https://github.com/kubernetes/kubernetes/pull/19758#issuecomment-172286935) here to make some inputs.
> @bgrant0607 should `Affinity` and the related types (like `NodeSelector`) go in v1 or in extensions/v1beta1? If we do the latter, then the v1 `PodSpec` would have a field that is a pointer to a type that lives in v1beta1. The situation we have here is a little different from when we added `LabelSelector`, since in that case all the types that used `LabelSelector` we put in v1beta1. In previous discussions we decided we didn't need to create a v1beta1.PodSpec, and would just AND the old nodeSelector map with the new Affinity stuff, so `PodSpec` will be in v1.

davidopp commented:
@k8s-bot test this please

bgrant0607 commented:
The way to introduce fields at a lower level of maturity than the overall API type is to use annotations.

(Even a new API version, like v2alpha1, would not solve this problem, since the resources are stored as v1 in etcd.)

davidopp commented:
I think we're confident enough in the API to consider Affinity to be "beta". Are you saying that because we want to use it from v1 PodSpec, it's not OK to put Affinity in v1beta1?

davidopp commented:
And if the answer is "yes" then it seems to imply that we will go directly from having Affinity in an annotation to being in v1?

hongchaodeng commented on plugin/pkg/scheduler/algorithm/priorities/node_affinity_test.go:
```diff
@@ -0,0 +1,133 @@
+
+import (
+	"reflect"
+	_ "sort"
```
remove unused imports

hongchaodeng commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -557,7 +557,180 @@ func TestPodFitsSelector(t *testing.T) {
+								},
+							},
+						},
+						SoftNodeAffinity: []api.SoftNodeAffinityTerm{
```
Why do we need SoftNodeAffinity in this test?

hongchaodeng commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -557,7 +557,180 @@ func TestPodFitsSelector(t *testing.T) {
+				"foo": "bar",
+			},
+			fits: true,
+			test: "same labels with Affinity with multiple NodeSelectTerms",
```
"node has the same labels as the pod Affinity..."

hongchaodeng commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -557,7 +557,180 @@ func TestPodFitsSelector(t *testing.T) {
+			labels: map[string]string{
+				"foo": "bar",
+			},
+			fits: false,
```
This doesn't sound right. Because
```
// An empty node selector term matches all objects. A null node selector term
// matches no objects.
type NodeSelectorTerm struct {
```

@davidopp This is serious problem..

kevin-wangzefeng replied:
No, currently we changed the semantic, but not description in the document and the code are not updated yet. We working on refreshing them.

hongchaodeng replied:
This is the comments of NodeSelectorTerm in api/types.go in this PR.
If it's changed, please change that as well.

hongchaodeng commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -557,7 +557,180 @@ func TestPodFitsSelector(t *testing.T) {
+			labels: map[string]string{
+				"foo": "bar",
+			},
+			fits: false,
```
This isn't defined in #18261..

davidopp commented:
@bgrant0607 helped clarify my thinking on the issue I raised above. IIUC, we can't add Affinity to v1beta1 since it is not an object, it is just a field in PodSpec, which is in v1. We definitely don't want to make Affinity be a separate kind. So our only choices are to make it part of v1 or make it an annotation. I am totally fine with putting it in v1beta1 if that were possible, but I don't think we should put something that's brand new directly into v1.

@kevin-wangzefeng how would you feel about making it an annotation in 1.2? I think we would have to promote it from there directly into v1 in 1.3, unless there is something I'm misunderstanding.

kevin-wangzefeng commented:
> @kevin-wangzefeng how would you feel about making it an annotation in 1.2? I think we would have to promote it from there directly into v1 in 1.3, unless there is something I'm misunderstanding.

do you mean we first make it an annotation in 1.2 and then promote it from the annotation form directly into v1 in 1.3?
I'm not very clear about how it looks like in an annotation form, can you give some explanation?

kevin-wangzefeng commented:
@davidopp 
I just thought about making the affinity an annotation, would it be like the form below?
```
"annotations": {
   "affinity/node/require-schedule-ignore-execute": "az in (az1,az2)"
}
```
(I add some "-" in "require-schedule-ignore-execute" just to help reading)

kevin-wangzefeng commented:
@davidopp Implementation refreshed with the annotation approach, PTAL

davidopp commented:
@kevin-wangzefeng How did embedding the JSON as a string in the annotation value work without any escaping? I thought from the discussion in #18670 and #12226 (and our in-person discussion with Brian) that it wouldn't work.

kevin-wangzefeng commented:
@davidopp To embed the JSON in the annotation value inside a JSON, it must be string with escaping. But as Brian mentioned in #12226, in yaml, we can simply use `' '` or `>` to escape the whole JSON string.
For example, in yaml, we do:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  labels:
    name: nginx
  annotations:
    scheduler.alpha.kubernetes.io/affinity: >
      {
        "nodeAffinity": {
          "requiredDuringSchedulingIgnoredDuringExecution": {
            "nodeSelectorTerms": [
              {
                "matchExpressions": [
                  {
                    "key": "az",
                    "operator": "In",
                    "values": ["az3", "az2"]
                  }
                ]
              }
            ]
          }
        }
      }
    another-annotation-key: "another-annotation-value"
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
```
And with the command `kubectl convert -f nginx-pod.yaml -o json`, we can get the one in json format.
```json
{
    "kind": "Pod",
    "apiVersion": "v1",
    "metadata": {
        "name": "nginx",
        "creationTimestamp": null,
        "labels": {
            "name": "nginx"
        },
        "annotations": {
            "another-annotation-key": "another-annotation-value",
            "scheduler.alpha.kubernetes.io/affinity": "{\n  \"nodeAffinity\": {\n    \"requiredDuringSchedulingIgnoredDuringExecution\": {\n      \"nodeSelectorTerms\": [\n        {\n          \"matchExpressions\": [\n            {\n              \"key\": \"az\",\n              \"operator\": \"In\",\n              \"values\": [\"az3\", \"az2\"]\n            }\n          ]\n        }\n      ]\n    }\n  }\n}\n"
        }
    },
    "spec": {
        "containers": [
            {
                "name": "nginx",
                "image": "nginx",
                "ports": [
                    {
                        "containerPort": 80,
                        "protocol": "TCP"
                    }
                ],
                "resources": {},
                "terminationMessagePath": "/dev/termination-log",
                "imagePullPolicy": "Always"
            }
        ],
        "restartPolicy": "Always",
        "terminationGracePeriodSeconds": 30,
        "dnsPolicy": "ClusterFirst",
        "securityContext": {}
    },
    "status": {}
}
```
I have to say, it looks ugly in json format, but in yaml format, it is acceptable.
And in this PR, the annotation approach is implemented with no extra requirements, user can run a pod directly with a yaml file as usual, no need to run the `kubectl convert` command if they don't want to.

davidopp commented:
What I meant was, for example in your e2e test in this PR, there is no `>` or `' '` in the value strings of the annotation. There are backticks, but I assume that is just to tell go that the string spans multiple lines. Do you not need the escaping because you create the pod directly? If so, can you write a test that verifies that a PodSpec with escaping and that uses the YAML approach works?

kevin-wangzefeng commented:
You are right, the current tests don't verify if a PodSpec with escaping and that uses the YAML approach works. I'm going to add a test to verify it.

davidopp commented:
Great, thanks!

When you write the test, you might want to use an example that you could put in the documentation for this feature, so you don't have to write a separate example for the documentation.

kevin-wangzefeng commented:
Got it.

kevin-wangzefeng commented:
@davidopp a new test added, PTAL

davidopp commented:
I think the API additions should go in pkg/api/types.go and pkg/api/v1/types.go, not in
plugin/pkg/scheduler/api/types.go

cc/ @bgrant0607

kevin-wangzefeng commented:
Do you mean to put the Types(NodeSeletor, NodeSelectorTerm etc.) in pkg/api/types.go and pkg/api/v1/types.go, but not add a new field to PodSpec?

davidopp commented:
You should add the Annotation field to PodSpec (in pkg/api/types.go and pkg/api/v1/types.go), and also put the definitions for the types that Annotation uses (NodeSelector, etc.) in pkg/api/types.go and pkg/api/v1/types.go. (I verified with @bgrant0607 earlier today.) There should be no changes to plugin/pkg/scheduler/api/types.go

kevin-wangzefeng commented:
> put the definitions for the types that Annotation uses (NodeSelector, etc.) in pkg/api/types.go and pkg/api/v1/types.go.

I'm clear with it, but have a question about:
> add the Annotation field to PodSpec (in pkg/api/types.go and pkg/api/v1/types.go)

Is it necessary? Currently there is already one field called Annotations in the api.ObjectMeta, which is included anonymously in Pod, so when accessing `pod.Annotations`, it actually goes into `pod.ObjectMeta.Annotations`. I don't think we should add another field with the name Annotations in PodSpec, or at least give it another name. Or we just use the existing one?

bgrant0607 commented:
@kevin-wangzefeng No new annotation field is required. I think @davidopp meant that the Affinity type should be added to those types.go files.

kevin-wangzefeng commented:
@bgrant0607 Alright, I will move the Affinity and related types to those types.go files, thanks.

davidopp commented:
Yes, sorry for the confusion, I meant what Brian said.

kevin-wangzefeng commented:
@davidopp  the Affinity type is moved to pkg/api, PTAL.

kevin-wangzefeng committed:
implement Node affinity and NodeSelector

kevin-wangzefeng commented:
@davidopp rebased, PTAL

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
 
+// A node selector represents the union of the results of one or more label queries
+// over a set of nodes; that is, it represents the OR of the selectors represented
+// by the nodeSelectorTerms.
```
s/nodeSelectorTerms/node selector terms/
also, mention that a null NodeSelector matches all objects, i.e. does not restrict scheduling

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+// over a set of nodes; that is, it represents the OR of the selectors represented
+// by the nodeSelectorTerms.
+type NodeSelector struct {
+	// nodeSelectorTerms is a list of node selector terms. The terms are ORed.
```
you can omit the JSON field name in all of the field descriptions. So for example this can just be "a list of node selector terms. The terms are ORed." I think that looks simpler. If you agree, please apply that suggestoin to all of the field comments.

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+	NodeSelectorOpLt           NodeSelectorOperator = "Lt"
+)
+
+// An Affinity is a group of affinity scheduling requirements,
```
s/An Affinity/Affinity/

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+)
+
+// An Affinity is a group of affinity scheduling requirements,
+// including node affinity and inter pod affinity.
```
Omit this line of the comment. This PR doesn't have pod affinity, and even when you add it, it will be obvious enough from looking at the struct that it has node affinity and inter-pod affinity.

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+// An Affinity is a group of affinity scheduling requirements,
+// including node affinity and inter pod affinity.
+type Affinity struct {
+	// NodeAffinity is a group of node affinity scheduling requirements.
```
I would change this to something like "describes node affinity scheduling requirements for the pod."

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+	NodeAffinity *NodeAffinity `json:"nodeAffinity,omitempty"`
+}
+
+// An NodeAffinity is a group of node affinity scheduling requirements.
```
s/An NodeAffinity/Node affinity/

davidopp replied:
Add a comment that if RequiredDuringSchedulingRequiredDuringExecution and RequiredDuringSchedulingIgnoredDuringExecution are both set, then both node selectors must be satisfied.

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+	// a node that violates one or more of the expressions. The node that is
+	// most preferred is the one with the greatest sum of weights, i.e.
+	// for each node that meets all of the scheduling requirements (resource
+	// request, RequiredDuringScheduling affinity expressions, etc.),
```
s/RequiredDuringScheduling/requiredDuringScheduling/

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+	// for each node that meets all of the scheduling requirements (resource
+	// request, RequiredDuringScheduling affinity expressions, etc.),
+	// compute a sum by iterating through the elements of this field and adding
+	// "weight" to the sum if the node matches the corresponding MatchExpressions; the
```
s/MatchExpressions/matchExpressions/

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+}
+
+// An empty preferred scheduling term matches all objects with implicit weight 0
+// (i.e. it's a no-op). A null preferred scheduling term matches no objects.
```
s/no objects/no objects (i.e. is also a no-op)/

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+// An empty preferred scheduling term matches all objects with implicit weight 0
+// (i.e. it's a no-op). A null preferred scheduling term matches no objects.
+type PreferredSchedulingTerm struct {
+	// weight is in the range 1-100
```
I would change this to "weight associated with matching the corresponding nodeSelectorTerm, in the range 1-100"

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+	// weight is in the range 1-100
+	Weight int `json:"weight"`
+	// matchExpressions is a list of node selector requirements. The requirements are ANDed.
+	MatchExpressions []NodeSelectorRequirement `json:"matchExpressions,omitempty"`
```
I know this is what we had in the design doc, but I think *NodeSelectorTerm would be simpler (my reason for doing []NodeSelectorRequirement makes no sense in retrospect.)
Then the comment would be "a node selector term, associated with the corresponding weight"

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,95 @@ const (
+	MatchExpressions []NodeSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// AffinityAnnotationKey represents the key of affinity data(json serialized)
```
I think we don't want this defined in an API file, but I emailed @bgrant0607 to make sure.

davidopp commented:
I think we need to add some validation code in pkg/api/validation/validation.go for the new types
I think the things you'd want to check are:
* NodeAffinity is not completely empty (i.e. at least one pointer is non-nil, or the slice is length>0)
* non-nil *NodeSelectorTerm in PreferredSchedulingTerm
* non-zero-length NodeSelector.NodeSelectorTerms
* adapt ValidateLabelSelector() from here https://github.com/kubernetes/kubernetes/blob/cf8d05f420c0b00085fa8385665327a3b44efb17/pkg/apis/extensions/validation/validation.go#L591
to use for validating NodeSelectorTerm

Does that sound reasonable?

kevin-wangzefeng commented:
@davidopp 
I'd like to add some validation code too, just have some issue on the package importing.
1. The validation functions are supposed to be called at run time when parsing affinity from serialized data in Pod.Annotations.
2. I think a proper way to do is call validation inside `GetAffinityFromPod()`. 
3. To avoid import cycle, better move `GetAffinityFromPod()` to somewhere else outside of `pkg/api` (helpers.go is part of `pkg/api`). validation imports pkg/api already, while it needs importing `pkg/api/validation` to call the validation functions.

Now I'm thinking to move `GetAffinityFromPod()` and `AffinityAnnodationKey` to scheduler (maybe a new helpers.go in `plugin/pkg/scheduler/algorithm`?), and then add the validations as you mentioned.

What do you think?

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -331,12 +331,66 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
 	return selector.PodSelectorMatches
 }
 
+// The pod can only schedule onto nodes that satisfy both required requirements in NodeAffinity and nodeSelector.
```
s/both required requirements in/requirements in both/

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -331,12 +331,66 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
-	if len(pod.Spec.NodeSelector) == 0 {
-		return true
+	// Check if node.Labels match pod.Spec.NodeSelector.
+	nodeSelectorMatches := true
```
You can simplify all of this logic (lines 337-344) to just

```
if len(pod.Spec.NodeSelector) > 0 {
  selector := ...
  if (!selector.Matches(labels.Set(nodeLabels)) {
     return false
   }
}
```

davidopp commented:
I see. Yeah, I guess moving GetAffinityFromPod() and AffinityAnnodationKey to scheduler is the best thing to do. It has two drawbacks though
* When affinity becomes a field (rather than annotation), we'll want the validation in the "normal" place, so we'll have to move it
* Also, when you implement kubelet admission and killing, it would be nice to be able to call GetAffinityFromPod(), but instead you will have to copy-and-paste. I guess this is OK, and when we move affinity to a field (and we move the validation to the normal place) we can have kubelet use the function directly.

kevin-wangzefeng commented:
Alright, I will them to scheduler. The bullet points are OK to me.

davidopp commented:
The goal of validation is to reject the creation request before the object is persisted. So I think you need to integrate your checks with the regular validation code (i.e. the code in validation.go). It is not useful to validate after the Pod object has been persisted (there is nothing useful you could do if you find a violation).

Sorry, I'm too tired right now to think through this clearly, but wanted to make that observation.

I will continue reviewing your PR tomorrow, but hopefully you have enough comments to work with for now.

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/node_affinity.go:
```diff
@@ -0,0 +1,97 @@
+		}
+		result = append(result, schedulerapi.HostPriority{Host: node.Name, Score: int(fScore)})
+		glog.V(10).Infof(
+			"%v -> %v: NodeAffinityPriority, Score: (%d)", pod.Name, node.Name, int(fScore),
```
unnecessary comma at end of line

davidopp commented on plugin/pkg/scheduler/algorithm/priorities/node_affinity.go:
```diff
@@ -0,0 +1,97 @@
+
+	result := []schedulerapi.HostPriority{}
+	for _, node := range nodes.Items {
+		fScore := float32(0)
```
probably better to use float64 everywhere here, since it is more accurate

davidopp commented:
It probably doesn't make sense to do validation for JSON embedded in an annotation. Maybe just skip it.

timothysc commented on docs/user-guide/node-selection/pod-with-node-affinity.yaml:
```diff
@@ -0,0 +1,28 @@
+apiVersion: v1
```
It's kind of weird to have an example without the complete text documentation to go along with it.  Fine for a follow PR if needed.

kevin-wangzefeng replied:
Yes, I will send follow-up PR to update the document.

kevin-wangzefeng commented:
> It probably doesn't make sense to do validation for JSON embedded in an annotation. Maybe just skip it.

Just to be clear, do you mean we skip validation, or we do the validation at run time?
With user's perspective, I think both approaches act the same:
1. User can create a pod with an invalid/misconfigured affinity in annotations without getting any error info.
2. The invalid/misconfigured affinity in pod.Annotations will sure make the pod unschedulable at scheduling time.

davidopp commented:
I think there is no import cycle if you implement the validation in ValidateAnnotations() (which is in pkg/api/validation/validation.go) and don't move GetAffinityFromPod().

Then validation.ValidateAnnotations() would call api.GetAffinityFromPod().

Does that work?

davidopp commented on pkg/api/helpers.go:
```diff
@@ -263,3 +264,50 @@ func ParseRFC3339(s string, nowFn func() unversioned.Time) (unversioned.Time, er
+// NodeSelectorRequirementsAsSelector converts the []NodeSelectorRequirement api type into a struct that implements
+// labels.Selector
+func NodeSelectorRequirementsAsSelector(nsm []NodeSelectorRequirement) (labels.Selector, error) {
+	if nsm == nil || len(nsm) == 0 {
```
len() for a nil slice returns 0, so you can drop the "if nsm == nill" part

kevin-wangzefeng commented:
That sounds better. 
And one concern left is that `ValidateAnnotations()` is kind of generally called function since it's called by `ValidateObjectMeta()`(whom almost everybody calls).
Another choice would be: implement the validation in `ValidatePod()` and `ValidatePodTemplateSpec()`, where the same level `ValidateAnnotations()` and `ValidatePodSpec()` are called.
Maybe it's just my uncalled-for worries, even if we do it in `ValidateAnnotations()`, the validation of affinity will be sure skipped if no value indicated.

davidopp commented:
That's a good point. But I think doing the validation from `ValidatePod()` and `ValidatePodUpdate()` is sufficient. Why would you do it in `ValidatePodTemplateSpec()`?

kevin-wangzefeng commented:
`ValidatePodTemplateSpec()` is called in the validation for RC.

kevin-wangzefeng commented:
The call chain is `ValidateReplicationControllerSpec()` -> `ValidatePodTemplateSpecForRC()` -> `ValidatePodTemplateSpec()`. `ValidatePodTemplateSpec()` will call `ValidateAnnotations()` and `ValidatePodSpec()` directly.

That is, if we want to put e.g. `ValidateAffinityInPodAnnotation()` at the same level with `ValidateAnnotations()` and `ValidatePodSpec()` called, we need to put it everywhere the `ValidateAnnotations()` is called for validating Pod.Annotations.

davidopp commented:
I see. I think it would be simpler to go back to my original suggestion of putting the affinity validation code in `ValidateAnnotations()` (or a function called from `ValidateAnnotations()`). You raise a good point that we will call it unnecessarily (e.g. for non-Pod objects), but it should incur very little overhead since those objects will not have an Affinity annotation. Ideally we would raise a validation error if we find Affinity annotation on non-Pod object (assuming `ValidateAnnotations()` knows the kind of the object -- if not, I think it's fine not to do this check).

Does that sound reasonable?

kevin-wangzefeng commented:
Yes, is much simpler to implement the validation in `ValidateAnnotations()`, I will go this way, thanks.

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -331,12 +331,66 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+
+	// 1. nil NodeSelector matches all nodes (i.e. does not filter out any nodes)
+	// 2. nil []NodeSelectorTerm (equivalent to non-nil empty NodeSelector) matches no nodes
+	// 3. zero-length non-nil []NodeSelectorTerm we can say matches no nodes also, just for simplicity
```
s/we can say//

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -331,12 +331,66 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+	// 2. nil []NodeSelectorTerm (equivalent to non-nil empty NodeSelector) matches no nodes
+	// 3. zero-length non-nil []NodeSelectorTerm we can say matches no nodes also, just for simplicity
+	// 4. nil []NodeSelectorRequirement (equivalent to non-nil empty NodeSelectorTerm) matches no nodes
+	// 5. zero-length non-nil []NodeSelectorRequirement we can say matches no nodes also, just for simplicity
```
s/we can say//

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -331,12 +331,66 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+			nodeSelectorTerms = affinity.NodeAffinity.RequiredDuringSchedulingRequiredDuringExecution.NodeSelectorTerms
+		}
+		if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+			nodeSelectorTerms = append(nodeSelectorTerms, affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms...)
```
We didn't talk about this, but I think what we want is "if RequiredDuringSchedulingRequiredDuringExecution and RequiredDuringSchedulingIgnoredDuringExecution are both set, then both node selectors must be satisfied." Thus you can't just append the two lists of nodeSelectorTerms, since the nodeSelectorTerms get ORed. You have to run through the "match node selector term by term" loop once for RequiredDuringSchedulingRequiredDuringExecution and once for RequiredDuringSchedulingIgnoredDuringExecution and only return true if both are satisfied.

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates_test.go:
```diff
@@ -575,7 +575,188 @@ func TestPodFitsSelector(t *testing.T) {
+			labels: map[string]string{
+				"foo": "bar",
+			},
+			fits: true,
```
this should be false; see my earlier comment about how to handle the situation where both are set

davidopp commented:
Please add a few more tests to predicate_test.go:
* something that uses Lt (or Gt)
* something that uses NotIn
* something that uses Exists (or DoesNotExist)
* something that uses multiple MatchExpression's in the same nodeSelectorTerm (to show that they are ANDed)

davidopp commented:
A few more:
* Pod with no Affinity will schedule onto a node
* Pod with Affinity but nil NodeSelector will schedule onto a node
* Pod with an Affinity and a PodSpec.NodeSelector (the old thing that we are deprecating): both must be satisfied for the pod to schedule onto the node

davidopp commented on test/e2e/scheduler_predicates.go:
```diff
@@ -470,4 +471,220 @@ var _ = Describe("SchedulerPredicates [Serial]", func() {
+				Annotations: map[string]string{
+					"scheduler.alpha.kubernetes.io/affinity": `
+						{"nodeAffinity": { "requiredDuringSchedulingIgnoredDuringExecution": {
+							"nodeSelectorTerms": [
```
If you were trying to test AND here, you need to put both matchExpressions them in the same nodeSelectorTerm

davidopp commented:
OK, I finally finished the review. Very nice job! Obviously I had a lot of comments, but I don't think any of them require big changes. Let me know when you've updated the PR and want me to take another look.

kevin-wangzefeng commented:
Thanks for your patiently review, I'm just done some of them, it won't take long to get rest of them done.

kevin-wangzefeng commented:
@davidopp Updated to address your comments, PTAL.
And I updated some description and adjusted the order of tests in `predicates_test.go`, trying to make them more clear, hope not take too much of your time to review.

davidopp commented on pkg/api/helpers.go:
```diff
@@ -263,3 +264,54 @@ func ParseRFC3339(s string, nowFn func() unversioned.Time) (unversioned.Time, er
+	return selector, nil
+}
+
+// AffinityAnnotationKey represents the key of affinity data(json serialized)
```
s/data(json/data (json/

davidopp commented on pkg/api/types.go:
```diff
@@ -1041,6 +1041,93 @@ const (
+type PreferredSchedulingTerm struct {
+	// Weight associated with matching the corresponding nodeSelectorTerm, in the range 1-100.
+	Weight int `json:"weight"`
+	// A node selector term, associated with the corresponding weight.
```
Sorry, my suggestion was actually change the field to be *NodeSelctorTerm, as well as the comment.
So thinking about this more, how about this? (Sorry, I realize this requires changing the validation code and the priority function and the tests a little...)
```
// (same comment as you have now)
type PreferredSchedulingTerm struct {
   // Weight associated with matching the corresponding preference, in the range 1-100
  Weight int `json:"weight"`
  // A preference, associated with the correspnoding weight
  Preference *NodeSelectorTerm `json:"preference,omitempty"`
}
```

kevin-wangzefeng replied:
I'm fine with it.

davidopp replied:
Actually, before you change this, let's ask Brian if this is allowed (having a pointer that is a required field). It may be that what we had originally is the only correct way to do this.

davidopp replied:
Actually I'm 90% sure the right way to do this is
```
Preference NodeSelectorTerm `json:"preference"`
```

kevin-wangzefeng replied:
Yeah, I think this way is better.

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -1346,6 +1350,87 @@ func ValidatePodSpec(spec *api.PodSpec, fldPath *field.Path) field.ErrorList {
+func ValidateNodeSelector(nodeSelector *api.NodeSelector, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for _, nst := range nodeSelector.NodeSelectorTerms {
```
Now that we have validation, I think maybe we should reject
len(NodeSelector.NodeSelectorTerms) == 0
and also reject
len(NodeSelectorTerm.MatchExpressions) == 0
and also reject
len(NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution) == 0
and also reject
len(PreferredSchedulingTerm.MatchExpressions) == 0

I think those situations would only happen when user makes an error in their configuration, so it would be good to return error to the user. I realize we have rules on how to handle those scenarios right now, and I think it's OK to leave those rules in the code (i.e. I would not change any of the code you have in scheduler predicate or priority functions), but we can return error during validation (so the lower levels of the code should never see those situations). 

If you do that, then I would also remove (from types.go) the comments "A null or empty node selector term matches no objects" and "An empty preferred scheduling term matches all objects with implicit weight 0 (i.e. it's a no-op). A null preferred scheduling term matches no objects (i.e. is also a no-op)."

Does that sound reasonable?

kevin-wangzefeng replied:
> len(NodeSelector.NodeSelectorTerms) == 0
and also reject
len(NodeSelectorTerm.MatchExpressions) == 0
and also reject
len(NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution) == 0
and also reject
len(PreferredSchedulingTerm.MatchExpressions) == 0

if we reject these situations in validation, I guess we should also remove the corresponding "omitempty" in json tag, since the fields are required and wouldn't be empty? And I had a look in the existing api, some fields that are required have the "omitempty" in json tag while some don't, is there any rule/preference I can refer to or follow?
And I had a look in the existing api, some required

davidopp replied:
This is a good question. I assume you are right, we should remove "omitempty" so the swagger documentation will say the field is required. But let's ask Brian. I'll send an email.

kevin-wangzefeng replied:
@davidopp I just find out we cannot simply reject the situation `len(NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution) == 0`
because it is allowed to be nil -- `NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution` is not configured.
Other situations are fine.

davidopp replied:
Ah yes, you're right.

Also note that
len(PreferredSchedulingTerm.MatchExpressions) == 0
is not needed if we change that field as described in the other comment.

davidopp commented on pkg/api/validation/validation.go:
```diff
@@ -97,6 +97,10 @@ func ValidateAnnotations(annotations map[string]string, fldPath *field.Path) fie
 		allErrs = append(allErrs, field.TooLong(fldPath, "", totalAnnotationSizeLimitB))
 	}
+
+	if annotations[api.AffinityAnnotationKey] != "" {
```
is there a difference between this and `if annotations[api.AffinityAnnotationKey] > 0` ?
Is one better than the other?

kevin-wangzefeng replied:
Do you mean `if len(annotations[api.AffinityAnnotationKey]) > 0`?
I guess there's no difference, just in different way.

davidopp replied:
Yeah, but never mind, what you have is fine (and I see you did it the same way in GetAffinityFromPodAnnotations()). So no need to change anything.

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -331,12 +331,67 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+		// Match node selector for requiredDuringSchedulingRequiredDuringExecution.
+		if nodeAffinity.RequiredDuringSchedulingRequiredDuringExecution != nil {
+			nodeSelectorTerms := nodeAffinity.RequiredDuringSchedulingRequiredDuringExecution.NodeSelectorTerms
+			glog.V(10).Infof("Match for node selector terms %+v", nodeSelectorTerms)
```
Even though in can be figured out from the line number, probably better to write "Match for RequiredDuringSchedulingRequiredDuringExecution node selector terms"

davidopp commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -331,12 +331,67 @@ func NewSelectorMatchPredicate(info NodeInfo) algorithm.FitPredicate {
+		// Match node selector for requiredDuringSchedulingRequiredDuringExecution.
+		if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+			nodeSelectorTerms := nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
+			glog.V(10).Infof("Match for node selector terms %+v", nodeSelectorTerms)
```
Even though in can be figured out from the line number, probably better to write "Match for RequiredDuringSchedulingIgnoredDuringExecution node selector terms"

davidopp commented on test/e2e/scheduler_predicates.go:
```diff
@@ -470,4 +471,217 @@ var _ = Describe("SchedulerPredicates [Serial]", func() {
 		expectNoError(err)
 		Expect(labelPod.Spec.NodeName).To(Equal(nodeName))
 	})
+
```
Please add one e2e test for validation (something that fails validation).

davidopp commented:
Finished re-review.

kevin-wangzefeng commented on pkg/api/validation/validation.go:
```diff
@@ -1346,6 +1350,87 @@ func ValidatePodSpec(spec *api.PodSpec, fldPath *field.Path) field.ErrorList {
+func ValidatePreferredSchedulingTerms(terms []api.PreferredSchedulingTerm, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for _, term := range terms {
```
I found actually the range index should not be ignored here, it needs to be used in generating the error msg, e.g. the weight is not valid:
```go
	allErrs = append(allErrs, field.Invalid(fldPath.Index(i).Child("weight"), term.Weight, "must be in the range 1-100"))
```

davidopp commented:
LGTM

davidopp commented:
Ah, needs rebase.

kevin-wangzefeng commented:
@davidopp rebased, PTAL.

davidopp commented:
Please squash the commits into a single commit.

kevin-wangzefeng commented:
@davidopp Sorry, squashed, PTAL.

davidopp commented:
@k8s-bot test this issue: #17873 #20402

davidopp commented:
LGTM

erictune commented:
@davidopp if the author is asleep, you should still be able to rebase.  DM me for instructions.

kevin-wangzefeng commented:
@davidopp Rebased, PTAL.

davidopp commented:
@k8s-bot unit test this issue: #20613

davidopp commented:
PR builder e2e and unit/integration tests are green. Merge queue is blocked. I'm merging this PR manually now because I'm OOO the rest of the week and the PR author is OOO next week.
